### PR TITLE
Remove root from cargo lock

### PR DESCRIPTION
**Pull request by raytung**
Removing [root] section from Cargo.lock for #4563 

Tests of interest has been updated accordingly, especially [tests/lockfile-compat.rs#oldest_lockfile_still_works](https://github.com/rust-lang/cargo/compare/master...raytung:remove-root-from-cargo-lock?expand=1#diff-48866d1891f97cb799dd0ca7148d1eefR10)

### CODE REVIEW

File: src/cargo/ops/lockfile.rs [hunk 1]
```diff
@@ -40,30 +40,14 @@ pub fn write_pkg_lockfile(ws: &Workspace, resolve: &Resolve) -> CargoResult<()>
...
-        !orig.starts_with("[[package]]")
-    } else {
-        true
-    };
-
     let toml = toml::Value::try_from(WorkspaceResolve {
         ws: ws,
         resolve: resolve,
-        use_root_key: use_root_key,
+        use_root_key: false,
```
Comment (matklad) [inline src/cargo/ops/lockfile.rs]: I think this field can be removed now as well?

File: tests/git.rs [hunk 2]
```diff
@@ -971,7 +971,7 @@ fn stale_cached_version() {
     let rev = repo.revparse_single("HEAD").unwrap().id();
 
     File::create(&foo.root().join("Cargo.lock")).unwrap().write_all(format!(r#"
-        [root]
+        [[package]]
```
Comment (matklad) [inline tests/git.rs]: Is it true that changes in tests are mostly cosmetics? That is, the tests pass with `[root]` as well as with `[[package]]`?

File: src/doc/guide.md [hunk 3]
```diff
@@ -328,7 +328,7 @@ Cargo will take the latest commit and write that information out into our
 `Cargo.lock` when we build for the first time. That file will look like this:
 
 ```toml
-[root]
+[[package]]
```
Comment (matklad) [inline src/doc/guide.md]: Excellent catch! I haven't thought of the docs at all!

**Comment by lukaslueg** [comment]
A workspace with an old-style lock file should get transformed by `cargo update`, shouldn't it? Can we have a test (extended) to see to that effect?

**Comment by matklad** [comment]
@lukaslueg yeah, I think it makes sense to add this test as well. `tests/lockfile-compat.rs` sounds like a proper place for it!

File: tests/git.rs (hunk 2 above)
Comment (raytung) [reply tests/git.rs]: Looks like it. I made the same change against master branch and the test passes without a hitch

File: tests/lockfile-compat.rs [hunk 4]
```diff
@@ -52,18 +52,25 @@ source = "registry+https://github.com/rust-lang/crates.io-index"
...
+    let cargo_commands = vec![
+        "build",
+        "update"
+    ];
 
-    let lock = p.read_lockfile();
-    for (l, r) in expected_lock_file.lines().zip(lock.lines()) {
-        assert!(lines_match(l, r), "Lines differ:\n{}\n\n{}", l, r);
-    }
+    for cargo_command in cargo_commands {
```
Comment (matklad) [inline tests/lockfile-compat.rs]: Hm, I think this won't work as expected, because the second time around the lockfile will already be updated.

Comment (raytung) [reply tests/lockfile-compat.rs]: Ah! of course. I'll probably create a helper method and loop through the vector.

**Comment by matklad** [comment]
LGTM!

I'll tag this relnotes: it's probably worth mentioning that lockfiles might loose `[root]` section on the next update.

@bors r+

**Comment by raytung** [comment]
Cheers @matklad and @lukaslueg! 😄

**Comment by budziq** [comment]
Hi I was recently hit by this https://github.com/budziq/rust-skeptic/issues/65. I'm parsing cargo.lock `[root]` section to get version numbers of  resolved dependencies (strangely this still works in nightly). 

What would be the correct way to get versions of all deps in the workspace?

**Comment by matklad** [comment]
@budziq `cargo metadata` is the way to go! http://doc.crates.io/external-tools.html#information-about-project-structure

**Comment by budziq** [comment]
Thanks @matklad! Is it's output stable?

**Comment by matklad** [comment]
@alexcrichton sometimes I think we should add a 

```
# The format of this file is unspecified and may change between Cargo releases.
# Use `cargo metadata` command to get the information about resolved dependencies in a stable way.
```

comment to the top of our lockfiles ... :)

**Comment by matklad** [comment]
@budziq yep! Moreover, it'll yell at you if you don't pass the `--format-version 1` argument explicitly :)

**Comment by budziq** [comment]
Thanks!

**Pull request merged**